File: desktop/agent/projectstore.go
package main

import (
	"context"
	"errors"
	"fmt"
	"io"
)

// projectstore.go defines the unified Read/Write contract that every
// runtime tier of a phone project must satisfy. Today three storages
// hold a project:
//
//   • the agent's SQLite-backed runtime (~/.yaver/phone-projects/<slug>/)
//   • a developer's git repo (.yaver/ directory inside the repo)
//   • the phone's offline sandbox (expo-sqlite on-device)
//
// All three reduce to two primitives: Read(slug) → Project and
// Write(Project) → ProjectMeta. Every higher-level verb (push, pull,
// export, import, deploy) is sugar on top of these primitives plus
// an HTTP transport when the source and destination live in
// different processes.
//
// Self-hosted and Yaver Cloud are the same tier — both run this same
// agent binary — so the agent has a single ProjectStore implementation
// (AgentProjectStore, defined alongside phone_backend.go) that backs
// both of them. The repo and phone-sandbox tiers each get their own
// implementation.
//
// This file declares the contract only. Implementations land in
// follow-up commits per the slice plan in
// docs/yaver-code-deploy-integration.md.

// ProjectStore is the unified Read/Write contract for a phone project
// across the three runtime tiers. Implementations MUST keep the
// canonical project files (schema, auth, seed, app spec) in sync with
// whatever lower-level storage they back onto, and MUST surface
// ProjectMeta deterministically so callers can reason about where a
// project currently lives without re-reading the whole bundle.
type ProjectStore interface {
	// List returns metadata for every project this store knows about.
	// The slice MAY be empty but MUST NOT be nil.
	List(ctx context.Context) ([]ProjectMeta, error)

	// Read materialises a project's canonical files into memory.
	// Returns ErrProjectNotFound if the slug is unknown to this store.
	// The returned Project's Stats field MAY be nil — stats are an
	// agent-only concern; the repo store has nothing to report.
	Read(ctx context.Context, slug string) (Project, error)

	// Write commits a project bundle into the store. WriteOptions
	// controls conflict behaviour when the slug already exists.
	// The returned ProjectMeta reflects the resolved slug (which may
	// differ from p.Slug under ConflictRename) and the post-write
	// timestamps so the caller doesn't have to re-Read.
	Write(ctx context.Context, p Project, opts WriteOptions) (ProjectMeta, error)

	// Snapshot dumps the live row data for the project, optionally
	// filtered by table or capped by row count. The agent store
	// drives this from the SQLite backend; the repo store reads from
	// .yaver/snapshots/. Phone-sandbox reads from on-device SQLite.
	// Returns ErrProjectNotFound if the slug is unknown.
	Snapshot(ctx context.Context, slug string, opts SnapshotOptions) (Snapshot, error)

	// ApplySnapshot writes a snapshot back into the store's live data
	// using the same conflict semantics as Write (controlled by the
	// snapshot's onConflict field). A no-op for stores that only hold
	// the declarative bundle without live rows.
	ApplySnapshot(ctx context.Context, slug string, snap Snapshot) error
}

// ProjectMeta is the small struct callers consult before deciding
// whether to do a full Read. It mirrors the fields a list endpoint
// would surface.
type ProjectMeta struct {
	Slug      string `json:"slug"`
	Name      string `json:"name"`
	Template  string `json:"template,omitempty"`
	CreatedAt string `json:"createdAt,omitempty"`
	UpdatedAt string `json:"updatedAt,omitempty"`
	// Tier identifies which storage this project lives in. One of
	// "agent", "repo", "phone-sandbox". Useful when callers compose
	// multiple stores.
	Tier string `json:"tier,omitempty"`
}

// Project is the canonical in-memory representation of a phone
// project. The fields mirror the existing PhoneProject value used by
// phone_backend.go but are restated here so a future package split
// (e.g. moving projectstore into its own Go module) does not pull in
// SQLite or HTTP machinery.
type Project struct {
	Slug      string
	Name      string
	Template  string
	CreatedAt string
	UpdatedAt string
	Schema    *PhoneSchema
	Auth      *PhoneAuth
	Seed      PhoneSeed
	App       *PhoneAppSpec
	// Stats are populated by the agent store from the SQLite backend.
	// Repo and phone-sandbox stores leave this nil.
	Stats *PhoneStats
	// Targets records the deployment bindings the project knows about
	// (dev-hw / yaver-cloud / cloudflare-workers). Persisted to
	// .yaver/project.yaml in the repo tier.
	Targets []TargetBind
	// TokenLabels is the list of token *labels* (no secret material).
	// Persisted to .yaver/tokens.lock.yaml. The actual hashed tokens
	// live per-machine in ~/.yaver/phone-projects/<slug>/tokens.yaml
	// and never leave the machine that minted them.
	TokenLabels []TokenLabel
}

// TargetBind records a single deploy binding. Unbound targets are
// omitted entirely; an empty slice means "no targets bound yet."
type TargetBind struct {
	Kind     string `yaml:"kind" json:"kind"` // dev-hw | yaver-cloud | cloudflare-workers
	BaseURL  string `yaml:"baseUrl,omitempty" json:"baseUrl,omitempty"`
	LastSync string `yaml:"lastSync,omitempty" json:"lastSync,omitempty"`
	Notes    string `yaml:"notes,omitempty" json:"notes,omitempty"`
}

// TokenLabel is the public-safe metadata for an API token. The
// actual `pp_<slug>_<hex>` secret never appears here.
type TokenLabel struct {
	ID     string   `yaml:"id" json:"id"`
	Label  string   `yaml:"label" json:"label"`
	Scopes []string `yaml:"scopes,omitempty" json:"scopes,omitempty"`
	CORS   []string `yaml:"cors,omitempty" json:"cors,omitempty"`
}

// WriteOptions tunes the Write call. Defaults reject collisions —
// ConflictPolicy must be set explicitly to overwrite or rename, the
// same way the existing /phone/projects/receive endpoint behaves.
type WriteOptions struct {
	OnConflict ConflictPolicy
	// SkipSeed leaves Project.Seed in memory but does not replay it
	// onto the live runtime. Lets callers replay seed only on first
	// import, not on every push.
	SkipSeed bool
	// IncludeData, when true, expects Project to carry a non-nil
	// Stats and an inline snapshot (or for the underlying transport
	// to follow up with ApplySnapshot). When false, only the
	// declarative bundle is written.
	IncludeData bool
}

// ConflictPolicy mirrors the existing phone-receive contract.
type ConflictPolicy string

const (
	ConflictReject    ConflictPolicy = ""          // default — fail if slug exists
	ConflictRename    ConflictPolicy = "rename"    // suffix the slug to make it unique
	ConflictOverwrite ConflictPolicy = "overwrite" // replace the existing project
)

// SnapshotOptions tunes Snapshot.
type SnapshotOptions struct {
	// Tables, when non-empty, limits the dump to the named tables.
	// Empty means "every table in the schema."
	Tables []string
	// MaxRowsPerTable caps the rows per table. Zero means unbounded.
	MaxRowsPerTable int
	// Compress requests gzip framing on the underlying io.Reader the
	// caller will read. Implementations that don't compress treat
	// this as a hint; HTTP transports use it to set Content-Encoding.
	Compress bool
}

// Snapshot is a streaming dump of live data. Body is JSONL by
// convention (one row per line, prefixed with the table name). The
// caller is responsible for closing Body.
type Snapshot struct {
	Slug       string
	TakenAt    string
	Rows       int64
	OnConflict ConflictPolicy // how ApplySnapshot should resolve row collisions
	Body       io.ReadCloser
}

// ErrProjectNotFound is the canonical sentinel for "no project with
// this slug." HTTP transports translate this to 404; CLI surfaces
// translate it to a clear "no such project" message instead of a
// generic "internal error." Callers detect it with errors.Is.
var ErrProjectNotFound = errors.New("project not found")

// NewProjectNotFound returns an error that carries the slug for
// human-facing messages while still satisfying
// errors.Is(err, ErrProjectNotFound) so callers at the HTTP/CLI
// boundary can map to 404 / non-generic error text.
func NewProjectNotFound(slug string) error {
	return fmt.Errorf("%w: %s", ErrProjectNotFound, slug)
}

File: mobile/src/lib/projectStore.ts
// projectStore.ts — TS twin of desktop/agent/projectstore.go.
//
// Defines the ProjectStore contract that every runtime tier must
// satisfy and ships the agent-tier implementation (HTTP). The
// phone-sandbox-tier implementation lives in projectStoreSandbox.ts
// so it can pull in expo-sqlite without breaking the path-aliased
// mobile-headless build (which does not have a SQLite shim).
//
// Self-hosted and Yaver Cloud both run the same agent binary, so a
// single agentStore implementation backs both — the only thing that
// differs is the baseUrl. That mirrors the "self-hosted and cloud
// same" simplification documented in
// docs/yaver-code-deploy-integration.md.

import type {
  PhoneAppSpec,
  PhoneAuth,
  PhoneProject,
  PhoneSchema,
  PhoneSeed,
  PhoneStats,
} from "./phoneProjects";

/** Tier label every ProjectStore stamps onto its ProjectMeta. */
export type ProjectTier = "agent" | "repo" | "phone-sandbox";

/** Conflict policy mirrors the existing /phone/projects/receive
 *  contract on the agent side. Keep the strings stable — they go
 *  on the wire as form fields. */
export type ConflictPolicy = "" | "reject" | "rename" | "overwrite";

export interface ProjectMeta {
  slug: string;
  name: string;
  template?: string;
  createdAt?: string;
  updatedAt?: string;
  tier: ProjectTier;
}

/** Project is the canonical in-memory representation. Mirrors the Go
 *  struct from desktop/agent/projectstore.go field-for-field so a TS
 *  client can shuttle a project between tiers without losing data. */
export interface Project {
  slug: string;
  name: string;
  template?: string;
  createdAt?: string;
  updatedAt?: string;
  schema?: PhoneSchema | null;
  auth?: PhoneAuth | null;
  seed?: PhoneSeed | null;
  app?: PhoneAppSpec | null;
  /** Live counts — agent tier only. */
  stats?: PhoneStats | null;
}

export interface WriteOptions {
  onConflict?: ConflictPolicy;
  skipSeed?: boolean;
  /** Whether to ship live-data alongside the declarative bundle. */
  includeData?: boolean;
}

export interface ProjectStore {
  list(): Promise<ProjectMeta[]>;
  read(slug: string): Promise<Project>;
  write(p: Project, opts?: WriteOptions): Promise<ProjectMeta>;
}

/** ProjectNotFoundError is the canonical sentinel for "no project
 *  with this slug." HTTP transports translate to 404; UI surfaces
 *  translate to clear "no such project" copy. */
export class ProjectNotFoundError extends Error {
  readonly slug: string;
  constructor(slug: string) {
    super(`project not found: ${slug}`);
    this.name = "ProjectNotFoundError";
    this.slug = slug;
  }
}

export function isProjectNotFound(err: unknown): err is ProjectNotFoundError {
  return err instanceof ProjectNotFoundError;
}

// ---- Agent tier (HTTP) ----------------------------------------------

export interface AgentStoreOptions {
  /** Trailing slash optional; the store strips it. */
  baseUrl: string;
  /** Auth + relay headers — the caller is responsible for wiring
   *  Authorization, X-Relay-Password, etc. */
  headers?: Record<string, string>;
  /** Override the fetch implementation. Defaults to globalThis.fetch
   *  so the same code runs in RN, browser, Node, and Bun. */
  fetchImpl?: typeof fetch;
}

/** agentStore returns a ProjectStore that talks to a running yaver
 *  agent's /phone/projects/* endpoints. Both self-hosted and
 *  Yaver Cloud are reached the same way; only the baseUrl differs. */
export function agentStore(opts: AgentStoreOptions): ProjectStore {
  const base = opts.baseUrl.replace(/\/+$/, "");
  const f = opts.fetchImpl ?? fetch;
  const headers = opts.headers ?? {};

  async function fetchJSON(path: string, init?: RequestInit): Promise<unknown> {
    const res = await f(base + path, {
      ...init,
      headers: { ...headers, ...(init?.headers ?? {}) },
    });
    if (res.status === 404) {
      // Caller wraps this with ProjectNotFoundError once it knows the
      // slug. Don't throw the typed error here — list() also hits 404
      // when the agent is too old to expose /phone/projects/list.
      throw Object.assign(new Error(`HTTP 404: ${path}`), { status: 404 });
    }
    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new Error(`HTTP ${res.status} ${path}: ${body.slice(0, 200)}`);
    }
    return res.json();
  }

  return {
    async list(): Promise<ProjectMeta[]> {
      const body = (await fetchJSON("/phone/projects/list")) as { projects?: PhoneProject[] } | null;
      const arr = Array.isArray(body?.projects) ? body!.projects! : [];
      return arr.map((p) => ({
        slug: p.slug,
        name: p.name,
        template: p.template,
        createdAt: p.createdAt,
        updatedAt: p.updatedAt,
        tier: "agent",
      }));
    },

    async read(slug: string): Promise<Project> {
      let body: unknown;
      try {
        body = await fetchJSON(`/phone/projects/get?slug=${encodeURIComponent(slug)}`);
      } catch (e: any) {
        if (e?.status === 404) throw new ProjectNotFoundError(slug);
        throw e;
      }
      if (!body || typeof body !== "object") throw new ProjectNotFoundError(slug);
      const p = body as PhoneProject;
      return {
        slug: p.slug,
        name: p.name,
        template: p.template,
        createdAt: p.createdAt,
        updatedAt: p.updatedAt,
        schema: p.schema ?? null,
        auth: p.auth ?? null,
        seed: p.seed ?? null,
        app: p.app ?? null,
        stats: p.stats ?? null,
      };
    },

    async write(p: Project, opts?: WriteOptions): Promise<ProjectMeta> {
      // The agent's /phone/projects/create accepts a PhoneCreateSpec.
      // Map our Project into that shape.
      const spec = {
        slug: p.slug,
        name: p.name,
        template: p.template ?? "",
        schema: p.schema ?? undefined,
        auth: p.auth ?? undefined,
        seed: opts?.skipSeed ? undefined : p.seed ?? undefined,
        app: p.app ?? undefined,
      };
      // Conflict handling on the agent's create endpoint isn't as
      // rich as /phone/projects/receive — we use the latter when the
      // caller asks for non-default conflict semantics. For now,
      // POST /create is the simple path; the upcoming wireMobilePush
      // helper covers the receive-with-bundle flow separately.
      const body = (await fetchJSON("/phone/projects/create", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(spec),
      })) as PhoneProject;
      return {
        slug: body.slug,
        name: body.name,
        template: body.template,
        createdAt: body.createdAt,
        updatedAt: body.updatedAt,
        tier: "agent",
      };
    },
  };
}

// ---- Convenience: pull from agent into any other store --------------

/** pullFromAgent reads a project off a remote agent and writes it
 *  into a destination store (typically the phone sandbox). This is
 *  the missing direction the design doc calls out: today the phone
 *  is push-only; this primitive closes the loop. */
export async function pullFromAgent(
  slug: string,
  agentOpts: AgentStoreOptions,
  dest: ProjectStore,
  writeOpts?: WriteOptions,
): Promise<ProjectMeta> {
  const src = agentStore(agentOpts);
  const project = await src.read(slug);
  return dest.write(project, writeOpts);
}
